### PR TITLE
Feature/651 reset password feature

### PR DESCRIPTION
**Pull request by inkhey**
required #27 , #25 

- [X] Add 3 endpoints  for reset password feature: generate, check_token, modify password
- [X] Refactor user objet function to have some generic function for token/password: hash, validate_hash, validate_date.
- [X] add hash and datetime for reset password in user table in db + migration script
- [X] add properly link with email -> add new notification type : reset_password who send a reset_password link.
- [X] add tests for most of this.
- [ ] clarify with frontend the reset_password url for email.

### CODE REVIEW

File: backend/development.ini.sample [hunk 1]
```diff
@@ -101,6 +101,9 @@ auth_type = internal
 # ldap_group_enabled = False
 # User auth token validity in seconds (used to interfaces like web calendars)
 user.auth_token.validity = 604800
+# user reset_password token validity (default to 600 -> 5 minutes)
```
Comment (lebouquetin) [inline backend/development.ini.sample]: what is the unit ?! 600 seconds = 10 minutes, not 5

Comment (inkhey) [reply backend/development.ini.sample]: just an error. i fix that.

File: backend/development.ini.sample [hunk 2]
```diff
@@ -101,6 +101,9 @@ auth_type = internal
 # ldap_group_enabled = False
 # User auth token validity in seconds (used to interfaces like web calendars)
 user.auth_token.validity = 604800
+# user reset_password token validity (default to 600s -> 10 minutes)
+user.reset_password.validity = 600
```
Comment (buxx) [inline backend/development.ini.sample]: This value has been validated by @lebouquetin ? Some mail provider can be slow, maybe 10 minute is too less.

File: backend/tracim_backend/config.py [hunk 3]
```diff
@@ -197,6 +197,10 @@ def __init__(self, settings):
             'user.auth_token.validity',
             '604800',
         ))
+        self.USER_RESET_PASSWORD_TOKEN_VALIDITY = int(settings.get(
+            'user.reset_password.validity',
+            '600'
```
Comment (buxx) [inline backend/tracim_backend/config.py]: Same as `user.reset_password.validity` parameter default value.

File: backend/tracim_backend/lib/mail_notifier/notifier.py [hunk 4]
```diff
@@ -418,6 +419,72 @@ def notify_created_account(
...
+        ))
+        translator = Translator(self.config, default_lang=user.lang)
+        async_email_sender = EmailSender(
+            self.config,
+            self._smtp_config,
+            self.config.EMAIL_NOTIFICATION_ACTIVATED
+        )
+        subject = self.config.EMAIL_NOTIFICATION_RESET_PASSWORD_SUBJECT.replace(
+            EST.WEBSITE_TITLE,
+            self.config.WEBSITE_TITLE.__str__()
```
Comment (buxx) [inline backend/tracim_backend/lib/mail_notifier/notifier.py]: Prefer `str(xxx)` instead `xxx.__str__()`

File: backend/tracim_backend/lib/mail_notifier/notifier.py [hunk 5]
```diff
@@ -418,6 +419,72 @@ def notify_created_account(
...
+        message['To'] = formataddr((user.get_display_name(), user.email))
+
+        text_template_file_path = self.config.EMAIL_NOTIFICATION_RESET_PASSWORD_TEMPLATE_TEXT  # nopep8
+        html_template_file_path = self.config.EMAIL_NOTIFICATION_RESET_PASSWORD_TEMPLATE_HTML  # nopep8
+        # TODO - G.M - 2018-08-17 - Generate token
+        context = {
+            'user': user,
+            # TODO - G.M - 11-06-2018 - [emailTemplateURL] correct value for logo_url  # nopep8
+            'logo_url': get_email_logo_frontend_url(self.config),
+            # TODO - G.M - 11-06-2018 - [emailTemplateURL] correct value for login_url  # nopep8
```
Comment (buxx) [inline backend/tracim_backend/lib/mail_notifier/notifier.py]: What do you use as value for `login_url` if it is a TODO ?

File: backend/development.ini.sample (hunk 2 above)
Comment (inkhey) [reply backend/development.ini.sample]: No validation now. I agree 10 minutes is a bit low, 15 minutes maybe ?

File: backend/tracim_backend/lib/mail_notifier/notifier.py (hunk 5 above)
Comment (inkhey) [reply backend/tracim_backend/lib/mail_notifier/notifier.py]: value are correct now, todo are deprecated. logo_url will probably be changed but not now.

File: backend/development.ini.sample (hunk 2 above)
Comment (buxx) [reply backend/development.ini.sample]: I was thinking as something like some one hour or two. Take a talk with @PhilippeAccorsi about that.

**Pull request merged**